### PR TITLE
Exceptions handling redesigned

### PR DESCRIPTION
**Pull request by paterczm**
This is a breaking change (apis changed). Please don't merge, just say if you're ok merging it.

I think that exception handling in lightblue-client is inconsistent and confusing in general. There are 3 issues pointing at problems with exceptions: #169, #143 and #188. This PR tries to address all of them.

List of exceptions and their usage (in this PR):
* LightblueException, a common parent to all lightblue exceptions
* LightblueHttpClientException, for http exceptions (no response or status != 200)
* LightblueResponseException, for when the response is valid, but contains errors. That means you can get a LightblueResponse object with errors only from this exception (valid response with errors used not cause exception to be thrown)
* LightblueResponseParseException, for json2pojo conversion errors and invalid response structure

In addition to exception related changes, I removed LightblueApacheHttpTransport. It's not used and does not have unit tests.

### CODE REVIEW

**Comment by paterczm** [comment]
@ykoer, @bserdar, @alechenninger - I'd like to get your opinions on this.

**Comment by BVulaj** [comment]
If one can only ever get LightblueResponse errors from an Exception, does it still make sense for them to live in the global Response object?  I understand that changing that at all is quite a breaking change though.

File: core/src/main/java/com/redhat/lightblue/client/response/LightblueResponseException.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+ * @author ykoer
+ */
+public class LightblueResponseException extends LightblueException implements LightblueResponseErrorCodes {
+
+    private static final long serialVersionUID = 1L;
+
+    private LightblueResponse lightblueResponse;
+    private Set<String> errorCodes;
+
+    public LightblueResponseException(String message, Throwable cause) {
```
Comment (bserdar) [inline core/src/main/java/com/redhat/lightblue/client/response/LightblueResponseException.java]: Is this ctor still needed? This exception class should have the response.

File: core/src/main/java/com/redhat/lightblue/client/response/LightblueResponseException.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+ * See {@link LightblueResponseErrorCodes} for a list of possible errors.
+ * Example: client can call e.exists(ERR_REST_CRUD_REST_SAVE)
+ *
+ * @author ykoer
+ */
+public class LightblueResponseException extends LightblueException implements LightblueResponseErrorCodes {
+
+    private static final long serialVersionUID = 1L;
+
+    private LightblueResponse lightblueResponse;
```
Comment (bserdar) [inline core/src/main/java/com/redhat/lightblue/client/response/LightblueResponseException.java]: Can this ever be null?

**Comment by paterczm** [comment]
@BVulaj 
> If one can only ever get LightblueResponse errors from an Exception, does it still make sense for them to live in the global Response object? I understand that changing that at all is quite a breaking change though.

The LightblueResponseException contains the LightblueResponse which contains errors. When LightblueResponse is returned normally, the errors are always null - is that what you're referring to? We could have LightblueResponse without error fields and methods and LightblueErrorResponse, extending LightblueResponse, which would have those.

**Comment by BVulaj** [comment]
@paterczm 
> The LightblueResponseException contains the LightblueResponse which contains errors. When LightblueResponse is returned normally, the errors are always null - is that what you're referring to? We could have LightblueResponse without error fields and methods and LightblueErrorResponse, extending LightblueResponse, which would have those.

Yep, that is what I was getting at and suggesting.  If the error field is always going to be null unless wrapped in an exception, I think we should do away with it and extend LightblueResponse

File: core/src/main/java/com/redhat/lightblue/client/response/LightblueResponseException.java (hunk 1 above)
Comment (paterczm) [inline core/src/main/java/com/redhat/lightblue/client/response/LightblueResponseException.java]: True.

File: core/src/main/java/com/redhat/lightblue/client/response/LightblueResponseException.java (hunk 2 above)
Comment (paterczm) [inline core/src/main/java/com/redhat/lightblue/client/response/LightblueResponseException.java]: No.

**Comment by paterczm** [comment]
@BVulaj, @bserdar, I incorporated your suggestions (2 commits at the top).

**Comment by bserdar** [comment]
I am fine with this change, but maybe we should release the current, and
inc the version number before merging this?

**Comment by paterczm** [comment]
> I am fine with this change, but maybe we should release the current, and
inc the version number before merging this?

Yes, we could even increase the major version, to indicate it's a breaking change.

**Comment by jewzaam** [comment]
Should `LightblueResponseParseException` extend `LightblueResponseException`?  Or rename to `LightblueParseException`.

+1 to major version change, as this is a breaking change.  Per our standard [1] it must be a major update assuming the client falls into the "application" bucket.

[1] http://docs.lightblue.io/standards/versioning.html

**Comment by paterczm** [comment]
> Should LightblueResponseParseException extend LightblueResponseException? Or rename to LightblueParseException.

LightblueResponseParseException means there are no errors in the response, but parsing failed - either because there is a mismatch between model and metadata or there was a problem parsing the response (client - server mismatch?).

Renamed it to LightblueParseException.

**Comment by alechenninger** [comment]
I argued for @BVulaj 's suggestion already in #169 so +1 (among other options but that was one of them)

**Comment by ykoer** [comment]
There are two kind of problems which we need to address:
1) Business Exceptions, which are expected, but not preventable and therefore reasonable to recover
2) Technical Exceptions, which is everything else. I would throw RuntimeException for those.

The problem here is how do we wanna determine the business errors from so many possible lightblue errors?

**Comment by paterczm** [comment]
LightblueResponseException represents "business" errors, like constraint violation. You need to check error code on that exception for specific error though.

**Pull request merged**